File: src/test/java/net/mreunionlabs/tapgradle/generate/GenerateTaskTest.java
package net.mreunionlabs.tapgradle.generate;

import org.gradle.api.Project;
import org.gradle.testfixtures.ProjectBuilder;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import static org.testng.Assert.*;

/**
 * Created by abangkis on 6/15/2016.
 */
public class GenerateTaskTest {
    private GeneratePluginExtension ext;
    private GenerateTask genTask;

    @BeforeMethod
    public void setUp() throws Exception {

        Project project = ProjectBuilder.builder().build();
        genTask = project.getTasks().create("genStructure", GenerateTask.class);

        ext = new GeneratePluginExtension();
        ext.setPackageString("net.mreunionlabs.test");
        ext.setJavaDir("test/gen");
    }

    @Test
    public void testGenerate() throws Exception {
        genTask.generate();
    }

}

File: src/test/kotlin/net/mreunionlabs/tapgradle/generate/GenerateTaskTest.kt
package net.mreunionlabs.tapgradle.generate

import io.kotlintest.TestCase
import io.kotlintest.specs.StringSpec
import org.gradle.api.Project
import org.gradle.kotlin.dsl.findByType
import org.gradle.testfixtures.ProjectBuilder

// see https://guides.gradle.org/testing-gradle-plugins
class GenerateTaskTest : StringSpec() {
    private lateinit var project: Project
    private lateinit var ext: GeneratePluginExtension

    override fun beforeTest(testCase: TestCase) {
        project = ProjectBuilder.builder().build()
        ext = GeneratePluginExtension()
        ext.packageString = "net.mreunionlabs.test"
        ext.javaDir = "gen/java"
        ext.resDir = "gen/resources"
        ext.webAppDir = "gen/webapp"
        ext.webDisplayName = "Project Testing"

    }

    init {

//        "get GeneratePlugin should not return null" {
//            project.pluginManager.apply("net.mreunionlabs.tapgradle.generate")
//            project.plugins.getPlugin(GeneratePlugin::class.java) shouldNotBe null
//        }
//
//        // see https://docs.gradle.org/current/userguide/custom_plugins.html for understanding how extensions works
        "run generate structure should not fail" {
            val ext = project.extensions.findByType(GeneratePluginExtension::class)
            ext?.packageString = "net.mreunionlabs.test"
            ext?.javaDir = "gen/java"
            ext?.resDir = "gen/resources"
            ext?.webAppDir = "gen/webapp"
            ext?.webDisplayName = "Project Testing"
            //fixme why setting ext programatically not working?

            val structureTask = project.tasks.create("genStructure", GenerateStructureTask::class.java)
            structureTask.generate()
            val layoutTask = project.tasks.create("genLayout", GenerateLayoutTask::class.java)
            layoutTask.generate()
        }
    }
}
